openhtf/io/frontend/src/app/utils.ts

```diff
@@ -0,0 +1,213 @@
+}
+
+
+/** A Pipe for darkening status colors even more. **/
+@Pipe({ name: 'colorToSuperDark' })
+export class ColorToSuperDark extends MapPipe {
+  map = {
+    'blue lighten-1 white-text': 'blue darken-4 white-text',
+    'green lighten-1 white-text': 'green darken-4 white-text',
+    'red lighten-1 white-text': 'red darken-4 white-text',
+    'yellow darken-3 white-text': 'yellow darken-5 white-text',
+    'blue-grey lighten-3 white-text': 'blue-grey lighten-1 white-text'
+  };
+}
+
+
+/** A Pipe for turning outcome codes into materialize-css color classes.**/
+@Pipe({ name: 'outcomeToColor' })
+export class OutcomeToColor extends MapPipe {
+  map = {
+    'PASS': 'green-text text-lighten-4',
+    'FAIL': 'red-text text-lighten-4',
+    'UNSET': 'blue-grey-text text-lighten-4'
+  };
+}
+
+
+/** A Pipe for colorizing log entries based on logging level. **/
+@Pipe({name: 'loggingLevelToColor'})
+export class LoggingLevelToColor extends MapPipe {
+  map = {
+    0: 'blue-text text-lighten-1',   // NOTSET
+    10: 'blue-text text-lighten-1',   // DEBUG
+    20: 'blue-text text-lighten-1',   // INFO
+    30: 'yellow-text text-darken-4',  // WARNING
+    40: 'red-text text-lighten-1',    // ERROR
+    50: 'red-text text-lighten-1'     // CRITICAL
+  };
+}
+
+
+/** A pipe for turning an Object into something Angular2 can loop over. **/
+@Pipe({name: 'objectToArray'})
+export class ObjectToArray implements PipeTransform {
+  transform(input: any): any {
+    if (input == undefined || input == null) {
+      return []
+    }
+    let keys = Object.keys(input),
+        result = [];
+
+    keys.forEach(key => { result.push({ 'key': key, 'value': input[key] }); });
+    return result;
+  }
+}
+
+
+/** A pipe for reducing a collection of measurements to a pass/fail result. **/
+@Pipe({name: 'measToPassFail'})
+export class MeasToPassFail implements PipeTransform {
+  transform(measurements: any): string {
+    let result = 'PASS'
+    Object.keys(measurements).forEach(key => {
+      if (measurements[key].outcome != 'PASS') {
+        result = 'FAIL';
+      }
+    })
+    return result;
+  }
+}
+
+
+/** A pipe for getting the first element of an array that only might exist. **/
+@Pipe({name: 'firstEltOrBlank'})
+export class FirstEltOrBlank implements PipeTransform {
+  /**
+   * Return the first element of the array if possible, or the empty string.
+   */
+  transform(input: any): string {
+    if (input && input.length > 0) {
+      return input[0];
+    }
+    return '';
+  }
+}
+
+
+/** Generic countdown timer. **/
+export class Countdown {
+  intervalID: number;
+  lengthSeconds: number;
+  millisRemaining: number = 0;
+  onDone: Function;
+
+  /**
+   * Create a new Countdown for the given number of seconds.
+   * @param {number} seconds - Length of the countdown in seconds.
+   */
+  constructor(seconds: number) {
+    this.lengthSeconds = seconds;
+    this.onDone = function() { };
+  }
+
+  /**
+   * Reset the countdown.
+   */
+  reset() {
```

Comment:
It seems reset is never used except right after `constructor` and before `start`. Maybe combine it into one of those, preferably `constructor` so that `millisRemaining` is always defined.